### DISCUSSION TITLE
"Various" releases image downloads (Mid-term project)

### DISCUSSION THREAD

**Discussion by Lovecraft-github** [ghDaYuYu/foo_discogs#96] [Ideas]
With the latest version, "Various" releases are not downloading artist images

**Comment by ghuDaYuYu** [comment]
Please, provide some working VA release IDs together with the last working version, was it v1.0.17 ?

**Comment by Lovecraft-github** [comment]
Releases that fails looks like this:
https://www.discogs.com/es/release/13010465-Various-Stone-Free-A-Tribute-To-Jimi-Hendrix
https://www.discogs.com/es/release/224256-Various-Covered-In-Nails-A-Tribute-To-Nine-Inch-Nails


I had difficulties finding a version that work (In the past it would do it because I have the images in the old directory). I tried 15 old versions (a lot of reset config, and new auth), no one can download artist images from these releases.
At some point discogs auth stop working, said working in test but fail when trying to download images, maybe due to so many OAuth attempts....? Anyway, I couldn't find one that do the work.

Is only the "Various" album artist, others releases with more than one artist download images just fine (All variations, albums with plenty of different artists, multiple artist in one song, etc.)

The cover download works in all kind of releases.

**Comment by ghuDaYuYu** [comment]
I was a bit confused at first reading the opening issue, even thought I was pretty sure that VA artist artwork download never worked in the past. I also run some tests anyway on foo_discogs and former foo_discogger versions to confirm it.
Implementation is a bit of a challenge as the component is quite album oriented, but will be nice to have it working in a future component release. 
All in all, should we move this issue to the discussion area (as a feature request) and continue the discussion there or maybe just close it?

**Comment by Lovecraft-github** [comment]
And now I'm confused, too.
For example
In [Various – From Dusk Till Dawn: Music From The Motion Picture](https://www.discogs.com/es/master/74278-Various-From-Dusk-Till-Dawn-Music-From-The-Motion-Picture)
The track 1 is credited to [George Clooney](https://www.discogs.com/es/artist/355356).
It is the only track that I have with DISCOGS_ARTIST_ID 355356 in my Library.
And I in my old "artist_art" folder I have the artist art from Clooney in a [folder named "355356"](https://imgur.com/USo7YQh) with the same images of [Clooney art](https://www.discogs.com/es/artist/355356-George-Clooney/images)
In [Various-Re-Covered-In-Nails-2001-A-Tribute-To-Nine-Inch-Nails](https://www.discogs.com/es/release/517295-Various-Re-Covered-In-Nails-2001-A-Tribute-To-Nine-Inch-Nails) the artist [Synical](https://www.discogs.com/es/artist/239295-Synical) meets the [same conditions](https://imgur.com/fMrhue4) 
[This folder](https://imgur.com/7QfZo7p) is the destination of foo_discogs (before) and foo_discogger (now). I never download (manually) any images in there, only the ones that the component download by itself.
These conditions (coming from one "Various" release and only present in my library in "Various artists" albums) are true for more artists in my collection, and I have folders with images named as its IDs.
I can't say to you when (I changed the disc in 03 July 2022 and old files have this date of creation), but some time in the past, the component had to download it.

Ok, I found one version that download the art images from "various" releases. [[v2.23-mod.09](https://github.com/ghDaYuYu/foo_discogs/releases/tag/v2.23-mod.09)]
[Here](https://imgur.com/TqLUpvY) and [here](https://imgur.com/60jwOJv)

**Comment by ghuDaYuYu** [comment]
Hi lovecraft, I am testing this issue just now.

Testing scenario...
- Foobar 2000 1.6.4 - plugin v2.23-mod.09
- Release (268443): https://www.discogs.com/release/268443-Various-Ghost-Dog-The-Way-Of-The-Samurai-The-Album
- Current selection: 18 tracks
- DISCOGS_ARTIST_ID tag in selection (18 tracks):
463631; 14747; 53764; 53765; 53766; 173160; 4848; 53768; 4244; 107494; 53769; 53771; 53772; 53773; 3081; 39655; 11213; 25495
- DISCOGS_MASTER_RELEASE_ID in selection: 85436
- DISCOGS_RELEASE_ID in selection : 268443
- Album artwork settings:  save to directory + not all files + overwrite existing
- Artist artwork settings: save to directory + not all files + overwrite existing
- Artist artwork setting 'Artist IDs to fetch for': %DISCOGS_ARTIST_ID% 
- No current artwork files in the selection folder
- Tag mapping: All disabled but DISCOGS_RELEASE_ID (write) and DISCOGS_ALBUM_ARTISTS (write + update).

Running 
WRITE TAGS...
- Find release panel opens with Artist, Filter and Release Id/Url text fields filled up. Artist list and release list empty.

NEXT BUTTON:
- Matching tracks panel pops up.
- All tracks match.

WRITE TAGS BUTTON:
- Download Album/Artist artwork multi task dialogs pops up.
- Track matching panel closes.

Test result:
- New tag DISCOGS_ALBUM_ARTISTS has been written (Various) to all tracks.
- 1 Image written to disk: cover.jpg

Caching preference content:
- Releases (1), Artists(22), Master Releases(1)

This is the result of this particular test, no artist artwork has been saved.

**Comment by ghuDaYuYu** [comment]
Also repeating same test scenario just aftwards, same result.

Console during test:
foo_discogs: https://i.discogs.com/.../rs:fit/g:sm/q:40/h:150/w:150/....jpeg
foo_discogs: https://i.discogs.com/.../rs:fit/g:sm/q:90/h:599/w:600/....jpeg

I am quite sure those 22 artists in the cache are not fully loaded.

It is not been until recent foo_discogger releases that multi-artist (not VA) support had been introduced, meaning that all artists are properly loaded (including artwork urls) by the time the write operation occurs.

Could you repeat the test scenario that downloaded VA artwork after clearing the plugin memory cache/restarting foobar?

**Comment by Lovecraft-github** [comment]
Sure, no problem
All new.
New wine prefix, clean portable installation of foobar2000_v1.6.16.exe
In library, only a folder with 18 random mp3s without tags.
Only non default component foo_discogs version: 2.23-mod.09 by DaYuyu Compiled: Mar 18 2021.
Configure as you.
Searching or entering the release number directly, I have an error pop-up "invalid stoi argument" and only a cover.jpg is downloaded.
Running again with cache gets the same result.
Cleaning cache, default settings in mapping, delete discogs tags in files.
Same result.
Running again with cache, no deleting tags; no pop up error, download the cover and Whitaker artist image.
Now that I have an artists_id tag, let's go try to update artist art. Download 12 artist images without error.
Cleaning the cache and updating artist art also gets 12 artist images.
Cleaning cache and restarting 12 images again.

In my default foobar. v2.0 beta 26
Cleaning cache and restarting.
Do the same test as before. Same results of v1.6.16

That makes me wonder, how can it download the first time I tried all images?

I started to changed %DISCOGS_ARTIST_ID% to %ARTIST_ID%,%ARTISTS_ID%.etc...No luck and one crashed foobar. I put back %DISCOGS_ARTIST_ID% and add to the path '\artist_art\' as i do in previous post, and it downloaded all the artists images like the time I tested the first time.
Then clean the cache to do it again, but this time only the Whitaker image. And I can reproduce it again.

It is all very strange....
The first time I installed the old discogs changing only the path I download two release with all images.
After cleaning the cache, the v2 and the 1.6 behaves consistently.
Then changing setting one time it downloaded all artists again, and after a new cleaning of the cache not work anymore.

I hope you could extract something from this, because I'm more confused than before.


Last edit, I hope:

I managed to reproduce it.

With the files tagged, update the artist art from the menu. Delete the images downloaded and do the normal write tags, and now it downloads all the images again.
I do this in v2 and 1.6 several times.
Changing the path or fetch wasn't related, the factor is using update at least one time for the release.
I supposed that old files were lying around in my profile from the time I used foo_discoggs.

Ufff, I started to think I was crazy..... :-P

Well, more info, when you restart foobar the trick didn't work anymore, You must have to update again for normal write tags to download images

**Comment by ghuDaYuYu** [comment]
Sure! haha...
Noticed this behavior while tagging different albums from related artists. So tagging a Judy Mowatt release first would provide her artwork when tagging The Wailers - Family and friends afterwards. It can also depend on the Searching preferences (auto-load releases on list selection), etc.

This was the reason to remove the option 'Artist to fetch for' (which was crashing from time to time) to start focusing on solving multi-artists first.

Unfortunately solving this issue, might take some time, and if properly done could lead to other new features, like embedding artist per track... for now I hacked the component just to do exactly that:

`1.0.17.2.Beta (VA trick)`

- Use with caution, or it will make a mess of you library organization, as it **trashes** %**RELEASE_ARTISTS_ID**% (**Album artist**) metadata, gets the job done though.
- Best used (**not for tagging**) with $directory_path(%path%)\%ARTIST_NAME% or similar option, and from '**Manage artwork**' ->  **Save artwork** button.

A proper solution will come to mind at some point, maybe by summer!


[foo_discogger mad 1.0.17.2 Beta.zip](https://github.com/ghDaYuYu/foo_discogs/files/10977653/foo_discogger.mad.1.0.17.2.Beta.zip)

**Comment by Lovecraft-github** [comment]
Luckily, I disabled auto-load (some times the first entry is not the correct one and I hate waiting to load, especially with artist with tons of releases, So I dodged that bullet.

No problem, I have plenty of released to move to the new artist image structure system. It is very possible that summer comes and I haven't finish.

I take a look at that release, the %ARTIST_NAME% is not a problem, in fact the new structure is to share images with biography script, and it has an internal BIO_ARTIST to structure the downloaded images, so it fits perfectly.

Anyway, the "normal" releases can have me busy till summer without problem.

Thanks for your time.

**Comment by ghuDaYuYu** [comment]
Your are welcome, thanks for taking the time to report, testing...
Till next time

**Comment by Lovecraft-github** [comment]
A pleasure.

**Comment by Lovecraft-github** [comment]
I saw that you start to implement it....

Cool
